deepinv/physics/generator/base.py

```diff
@@ -26,41 +29,75 @@ class PhysicsGenerator(nn.Module):
 
         >>> import torch
         >>> from deepinv.physics.generator import MotionBlurGenerator, SigmaGenerator
-        >>> _ = torch.manual_seed(0)
         >>> # combine a PhysicsGenerator for blur and noise level parameters
         >>> generator = MotionBlurGenerator(psf_size = (3, 3), num_channels = 1) + SigmaGenerator()
-        >>> params_dict = generator.step(batch_size=1) # dict_keys(['filter', 'sigma'])
+        >>> params_dict = generator.step(batch_size=1, seed=0) # dict_keys(['filter', 'sigma'])
         >>> print(params_dict['filter'])
         tensor([[[[0.0000, 0.1006, 0.0000],
                   [0.0000, 0.8994, 0.0000],
                   [0.0000, 0.0000, 0.0000]]]])
         >>> print(params_dict['sigma'])
-        tensor([0.1577])
+        tensor([0.2532])
     """
 
     def __init__(
-        self, step=lambda **kwargs: {}, device="cpu", dtype=torch.float32, **kwargs
+        self,
+        step=lambda **kwargs: {},
+        rng: torch.Generator = None,
+        device="cpu",
+        dtype=torch.float32,
+        **kwargs,
     ) -> None:
         super().__init__()
 
         self.step_func = step
         self.kwargs = kwargs
         self.factory_kwargs = {"device": device, "dtype": dtype}
+        self.device = device
+        if rng is None:
+            self.rng = torch.Generator(device=device)
+        else:
+            # Make sure that the random generator is on the same device as the physic generator
+            assert (
+                rng.device == device
+            ), "The random generator is not on the same device as the Physic Generator"
+            self.rng = rng
+        self.initial_random_state = self.rng.get_state()
```

Comment:
Can we have a `reset_rng` method that sets the rng back to the initial state? That way, we can have reproducibility in epochs for #202 (provided that `shuffle=False` in `DataLoader`). And add a test to test this.

Reply:
Good idea. I'll try to do this as soon as possible.

Reply:
```python
from torch.utils.data import DataLoader
import torch
from deepinv.tests.dummy_datasets.datasets import DummyCircles
from deepinv.physics import Inpainting
from deepinv.physics.generator import BernoulliSplittingMaskGenerator
from deepinv.utils import plot

# Sample dataset
img_size = (3, 8, 8)
dataset = DummyCircles(5, imsize=img_size)
loader = DataLoader(dataset, shuffle=False)

# Random inpainting physics
physics = Inpainting(tensor_size=img_size) # No mask for now
physics_generator = BernoulliSplittingMaskGenerator(tensor_size=img_size, split_ratio=0.5)

# Simulate a training loop for 2 epochs
out = []
for epoch in range(2):
    physics_generator.reset_rng()
    
    ys = []
    mask_prev = torch.tensor([0])
    for x in loader:
        params = physics_generator.step()
        y = physics(x, **params)
        ys += [y]
        # Test masks different every iteration
        assert not torch.equal(params["mask"], mask_prev)
        mask_prev = params["mask"]

    out += [torch.cat(ys, dim=-2)]

out = torch.cat(out, dim=-1)

# Test ys same every epoch
assert torch.equal(out[..., :8], out[..., 8:])

# Optional visualisation
plot(out)
```

Reply:
The issue #202 is solved now. To have the expected behavior, it's enough to call physiscs_generator.reset_rng() at the beginning of each epoch.

Reply:
Ok, I added the appropriate warning to close #202